File: backend/repositories/activity_repository.py
# backend/repositories/activity_repository.py
from typing import List, Dict
from repositories.base_repository import BaseRepository, DatabaseConnectionError, QueryError

ACTIVITY_TABLE = "ACT_PAR"


class ActivityRepositoryError(Exception):
    """Raised when the activity repository encounters an error."""
    pass


class ActivityRepository(BaseRepository):
    """Repository for activity data (table ACT_PAR)."""

    def get_all_activities(self) -> List[Dict]:
        """
        Return list of active activities from ACT_PAR table.

        Only returns activities where ACT_ACTF = 1 (active flag).
        Results are ordered by ACT_LIB (activity label).

        Returns:
            List[Dict]: List of dicts with keys: act_keyu, act_code, act_lib

        Raises:
            ActivityRepositoryError: on DB connection or query failure.
        """
        query = f"SELECT ACT_KEYU, ACT_CODE, ACT_LIB FROM {ACTIVITY_TABLE} WHERE ACT_ACTF = 1 ORDER BY ACT_LIB"
        try:
            rows = self.execute_query(query)
        except (DatabaseConnectionError, QueryError) as exc:
            raise ActivityRepositoryError(str(exc)) from exc

        # Convert rows (tuples) to list of dicts
        activities = []
        for row in rows:
            activities.append({
                "act_keyu": row[0],
                "act_code": row[1],
                "act_lib": row[2],
            })

        return activities

File: backend-go/repositories/activity_repository.go
package repositories

import (
	"database/sql"
	"fmt"
)

// ActivityRow is a single activity entry returned by the activity repository.
type ActivityRow struct {
	ActKeyu int
	ActCode string
	ActLib  string
}

// ActivityRepository provides access to activity data (table ACT_PAR).
type ActivityRepository struct {
	DB *sql.DB
}

// NewActivityRepository wires the shared connection pool into the repository.
func NewActivityRepository(db *sql.DB) *ActivityRepository {
	return &ActivityRepository{DB: db}
}

// GetAllActivities returns the active activities (ACT_ACTF = 1) ordered by label.
func (r *ActivityRepository) GetAllActivities() ([]ActivityRow, error) {
	const query = "SELECT ACT_KEYU, ACT_CODE, ACT_LIB FROM ACT_PAR WHERE ACT_ACTF = 1 ORDER BY ACT_LIB"

	rows, err := r.DB.Query(query)
	if err != nil {
		return nil, fmt.Errorf("get all activities failed: %w", err)
	}
	defer rows.Close()

	var activities []ActivityRow
	for rows.Next() {
		var activity ActivityRow
		if err := rows.Scan(&activity.ActKeyu, &activity.ActCode, &activity.ActLib); err != nil {
			return nil, fmt.Errorf("activity scan failed: %w", err)
		}
		activities = append(activities, activity)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("activity iteration failed: %w", err)
	}

	return activities, nil
}
